### DISCUSSION TITLE
A window with WindowChrome flickers during the resize operation if the content is WindowsFromsHost

### DISCUSSION THREAD

**Discussion by AndreyMarten** [dotnet/wpf#5897] [Q&A]
* .NET Core Version: 6.0
* Windows version: 21H2 (OS Build 22000.318)
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes

A window defined with **WindowChrome** flickers during the resize operation if the content is the **WindowsFromsHost** element. The screencast and project illustrate the problematic behavior. 
 
[Screencast](https://user-images.githubusercontent.com/31204429/147758949-f3313c8c-6dea-47d8-b5f7-a45596e528b2.mp4)
[sample.zip](https://github.com/dotnet/wpf/files/7793230/sample.zip)

The `WpfApp1` project is a .NET project that hosts a `UserControl1` from WinFormsLibary1 (also .NET). Please run this project to replicate the issue. 

The `DXSample` project is a .NET Framework project that hosts a `UserControl1` from WindowsFormsControlLibrary1 (also .NET Framework). In this project, you can see that the issue occurs in .NET Framework as well.

**Comment by maximilien-noal** [comment]
Windows Forms uses software rendering and it's going to flicker on resize in this context or if OnPaint() is slow enough,

Please use WPF instead. WPF uses hardware rendering and has therefore smooth rendering.

A very similar issue has been already labeled as "WON'T FIX' in the past :
https://social.msdn.microsoft.com/Forums/sqlserver/en-US/957d58bb-95f4-4bf0-9f87-aef55b564af8/windows-forms-host-in-wpf-application-flicker-on-visibility-change?forum=wpf

Quote :

> WPF and windows forms use different display technology.
>
>These things are oil and water.
>
>You could try reducing height and width of the containing control using an animation so you turn a problem into a feature. Just miss out the "from" and set the "To" to 1 or something and do 2 doubleanimations. One for height and one for width.
>
>Reduce the size then collapse visibility.

If you don"t want flickering, you have to use WPF. It has been made precisely to fix that problem (among others).

WindowsFormsHost is there so you can use existing 3rd paty WinForms controls or in-house WinForms controls that you'd like to use with WPF. It's meant to aid the transition to WPF. But if you can, only using WPF is the better choice.

This way, you'll also avoid the AirSpace problem :
https://github.com/dotnet/wpf/issues/152

**Comment by gurpreet-wpf** [comment]
@AndreyMarten I hope the answer provided by @maximilien-noal helped.

**Comment by batzen** [comment]
The "answer" provided is not an answer at all.

@AndreyMarten was explicitly stating that this happens when using `WindowChrome` and in fact using that causes excessive flickering.
Removing `WindowChrome` also removes the flickering.

> Windows Forms uses software rendering and it's going to flicker on resize in this context or if OnPaint() is slow enough,

That highly depends on double buffering etc.

> WPF uses hardware rendering and has therefore smooth rendering.

That's quite far away from reality as WPF does NOT always use hardware rendering.

> If you don"t want flickering, you have to use WPF. It has been made precisely to fix that problem (among others).

`WindowChrome` causes the issue, so using a WPF provided class causes the issue.

> WindowsFormsHost is there so you can use existing 3rd paty WinForms controls or in-house WinForms controls that you'd like to use with WPF. It's meant to aid the transition to WPF. But if you can, only using WPF is the better choice.

Try to use `WebBrowser` in the sample application and be surprised that, while a control provided by WPF, it also flickers.


@AndreyMarten Feel free to use the `WindowChromeBehavior` or even easier the `WindowChromeWindow` provided by ControlzEx. Your sample does not flicker when using that one.
You have to use the version from https://github.com/ControlzEx/ControlzEx/pull/151 which  will be part of ControlzEx 6.0.
The current version 5.0.1 of ControlzEx does not flicker in your sample but renders your control transparent when resizing, which is also problematic.
But the upcoming version makes your sample app work flawlessly.


@maximilien-noal My statements might sound a bit harsh, but to me it looks/sounds like you didn't even have a look at the provided sample application. It sounds like you read "WinForms" and made your judgement about what should and shouldn't be done. I don't think that's a good approach to reported issues.

@gurpreet-wpf May i ask why you converted this to a discussion without even waiting for a response from the reporter?

**Comment by maximilien-noal** [comment]
> That highly depends on double buffering etc.

Of course. I still remember how I had to fix flickering in the ListView by enabling DoubleBuffering, which is sadly a protected property and not enabled by default.

> Try to use WebBrowser in the sample application and be surprised that, while a control provided by WPF, it also flickers.

That's far from surprising, the WebBrowser content is not rendered by WPF.

> That's quite far away from reality as WPF does NOT always use hardware rendering.

That's true. However, hardware rendering is the norm in WPF land.

> WindowChrome causes the issue, so using a WPF provided class causes the issue.
> @AndreyMarten Feel free to use the WindowChromeBehavior or even easier the WindowChromeWindow provided by ControlzEx. Your sample does not flicker when using that one.
You have to use the version from ControlzEx/ControlzEx#151 which will be part of ControlzEx 6.0.
The current version 5.0.1 of ControlzEx does not flicker in your sample but renders your control transparent when resizing, which is also problematic.
But the upcoming version makes your sample app work flawlessly.

Interesting. Thank you !

**Comment by gurpreet-wpf** [comment]
Thank you @batzen for providing such detailed information. I didn't intend to close this issue without resolution. I will closely monitor the discussion to convert it back to issue, if there is a fix that can get into WPF.

I am happy to take your suggestion to wait for issue creator to respond.
